ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/druid/DruidQuery.java

```diff
@@ -0,0 +1,1053 @@
+      generator.writeStartObject();
+      generator.writeStringField("type", type);
+      generator.writeStringField("name", name);
+      writeFieldIf(generator, "fieldNames", fieldNames);
+      generator.writeEndObject();
+    }
+  }
+
+  /** Filter element of a Druid "groupBy" or "topN" query. */
+  private abstract static class JsonFilter implements Json {
+    final String type;
+
+    private JsonFilter(String type) {
+      this.type = type;
+    }
+  }
+
+  /** Equality filter. */
+  private static class JsonSelector extends JsonFilter {
+    private final String dimension;
+    private final String value;
+
+    private JsonSelector(String type, String dimension, String value) {
+      super(type);
+      this.dimension = dimension;
+      this.value = value;
+    }
+
+    public void write(JsonGenerator generator) throws IOException {
+      generator.writeStartObject();
+      generator.writeStringField("type", type);
+      generator.writeStringField("dimension", dimension);
+      generator.writeStringField("value", value);
+      generator.writeEndObject();
+    }
+  }
+
+  /** Bound filter. */
+  private static class JsonBound extends JsonFilter {
+    private final String dimension;
+    private final String lower;
+    private final boolean lowerStrict;
+    private final String upper;
+    private final boolean upperStrict;
+    private final boolean alphaNumeric;
+
+    private JsonBound(String type, String dimension, String lower,
+        boolean lowerStrict, String upper, boolean upperStrict,
+        boolean alphaNumeric) {
+      super(type);
+      this.dimension = dimension;
+      this.lower = lower;
+      this.lowerStrict = lowerStrict;
+      this.upper = upper;
+      this.upperStrict = upperStrict;
+      this.alphaNumeric = alphaNumeric;
+    }
+
+    public void write(JsonGenerator generator) throws IOException {
+      generator.writeStartObject();
+      generator.writeStringField("type", type);
+      generator.writeStringField("dimension", dimension);
+      if (lower != null) {
+        generator.writeStringField("lower", lower);
+        generator.writeBooleanField("lowerStrict", lowerStrict);
+      }
+      if (upper != null) {
+        generator.writeStringField("upper", upper);
+        generator.writeBooleanField("upperStrict", upperStrict);
+      }
+      generator.writeBooleanField("alphaNumeric", alphaNumeric);
+      generator.writeEndObject();
+    }
+  }
+
+  /** Filter that combines other filters using a boolean operator. */
+  private static class JsonCompositeFilter extends JsonFilter {
+    private final List<? extends JsonFilter> fields;
+
+    private JsonCompositeFilter(String type,
+        List<? extends JsonFilter> fields) {
+      super(type);
+      this.fields = fields;
+    }
+
+    public void write(JsonGenerator generator) throws IOException {
+      generator.writeStartObject();
+      generator.writeStringField("type", type);
+      switch ("type") {
```

Comment:
Ah, and now I see that I wrote that line before you copy-pasted it. Please fix it anyway!